### ISSUE TITLE
Better error messages on parse failure

### ISSUE THREAD

**Issue by rcoh** [rcoh/angle-grinder#12]

**Comment by tstack** [comment]
Did you have something in mind for how to handle this?  Looking at the nom error management docs, the [Merr pattern matching](https://github.com/Geal/nom/blob/master/doc/error_management.md#merr-pattern-matching) option seemed like a reasonable solution.

**Comment by rcoh** [comment]
I had no ideas about this -- at least when I was doing this awhile ago,
there didn't seem to be a straightforward solution

**Comment by rcoh** [comment]
Looked into this a bit more -- we probably want to combine early-return from errors with merr. (Although merr might be overkill, the grammar is pretty stable)

**Comment by tstack** [comment]
Cool.  I'm finishing up a change to add the `limit` operator.  I can try the different approaches for errors and see how it goes after I'm done with limit.

**Comment by rcoh** [comment]
Ah cool -- you may run into issues because currently non aggregate
operators are stateless currently (but there's no reason they need to be)

**Comment by tstack** [comment]
The "merr" stuff in nom doesn't seem to be working anymore.  The return_error!() macro seems to be sufficient, though.  I'll try to put together something that handles a couple error cases, like non-terminated string and no column names for `count by` and open a PR.

**Comment by tstack** [comment]
If you have some time, can you check out [this commit](https://github.com/tstack/angle-grinder/commit/6fac26aff9ca8692c8d49841b38045866298b59e) on my error-wip branch and say if you okay with the direction.

Here's a screenshot of the results so far:

![image](https://user-images.githubusercontent.com/107532/50784080-7c6a9c00-1261-11e9-9d4c-4e122fb1d75d.png)

**Comment by rcoh** [comment]
(this is from a fairly brief skim on my phone, more comments later)
First off, the results of **gorgeous**! I never would have thought to try to use the same library that rustc uses.

A couple of thoughts:
- I'm not in love with sticking postions ad-hoc into operators. What if we did something like:
```
WithPosition<T> {
  position: [Positional information needed for syntax errors]
  obj: T
}
```

We could have the parser return a seq of `Positioned<Operator>`

I think I'd prefer that strategy in general (wrapping objects with their positions) rather than having positional information leak into the underlying structs. We could even push that all the way down the execution engine for generating nice error messages at some point.

I'd also probably refactor the grammar a bit to push the positioning down a little bit so that you didn't need to create start and end tokens manually in most cases.

**Comment by rcoh** [comment]
eg.
```
Limit {
  count: Positioned<f64>
}
```

`WithPosition` can define `into` for `T` to easily extract the underlying type

**Comment by rcoh** [comment]
Another note: If it simplifies things, I'd be totally fine only exposing the first error in a pipeline

**Comment by tstack** [comment]
I don't like the position stuff either.  I like your proposal, I'll see if I can make it work.

**Comment by tstack** [comment]
I've updated my [error-wip branch](https://github.com/tstack/angle-grinder/commit/22ce9c2c37f2a506c21caa3e9ee4f5ca8389b246) if you want to take a look again.  Let me know if that's fine for a PR or if you want me to break it up or do some more changes.

I was able to get your Positioned suggestion working by adding a with_pos!() parser that wraps the result of the child parser with a Positioned object.  It's a definite improvement.

Thanks!
